File: com/craftinginterpreters/lox/Environment.java
package com.craftinginterpreters.lox;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Environment {
    static final Object UNINITIALIZED = new Object();
    final Environment enclosing;
    private final Map<String, Object> values = new HashMap<>();
    private final List<Object> valuesList = new ArrayList<>();

    Environment() {
        enclosing = null;
    }

    Environment(Environment enclosing) {
        this.enclosing = enclosing;
    }

    // Bind a variable name to a value in the environment.
    void define(String name, Object value) {
        values.put(name, value);
        valuesList.add(value);
    }

    // Walk up the chain of enclosing environments to find the one at the given distance.
    Environment ancestor(int distance) {
        Environment environment = this;
        for (int i = 0; i < distance; i++) {
            environment = environment.enclosing;
        }
        return environment;
    }

    // Fast O(1) indexed variable access for resolved local variables.
    Object getAt(int distance, int index) {
        Object value = ancestor(distance).valuesList.get(index);
        if (value == UNINITIALIZED) {
            throw new RuntimeError(new Token(TokenType.IDENTIFIER, "", null, -1),
                    "Variable used before initialization.");
        }
        return value;
    }

    // Fast O(1) indexed variable assignment for resolved local variables.
    void assignAt(int distance, int index, Object value) {
        ancestor(distance).valuesList.set(index, value);
    }

    // Get a variable from the environment at the given distance (by name).
    Object getAt(int distance, String name) {
        Object value = ancestor(distance).values.get(name);
        if (value == UNINITIALIZED) {
            throw new RuntimeError(new Token(TokenType.IDENTIFIER, name, null, -1),
                    "Variable '" + name + "' used before initialization.");
        }
        return value;
    }

    Object getAt(int distance, Token name) {
        Object value = ancestor(distance).values.get(name.lexeme);
        if (value == UNINITIALIZED) {
            throw new RuntimeError(name,
                    "Variable '" + name.lexeme + "' used before initialization.");
        }
        return value;
    }

    // Set a variable's value in the environment at the given distance (by name).
    void assignAt(int distance, Token name, Object value) {
        ancestor(distance).values.put(name.lexeme, value);
    }

    Object get(Token name) {
        if (values.containsKey(name.lexeme)) {
            Object value = values.get(name.lexeme);
            if (value == UNINITIALIZED) {
                throw new RuntimeError(name,
                        "Variable '" + name.lexeme + "' used before initialization.");
            }
            return value;
        }

        if (enclosing != null)
            return enclosing.get(name);

        throw new RuntimeError(name,
                "Undefined variable '" + name.lexeme + "'.");
    }

    // Assignment is not allowed to create new variables. It only updates existing
    // ones.
    void assign(Token name, Object value) {
        if (values.containsKey(name.lexeme)) {
            values.put(name.lexeme, value);
            return;
        }

        // If the variable isn't in this environment, try the enclosing one.
        if (enclosing != null) {
            enclosing.assign(name, value);
            return;
        }

        throw new RuntimeError(name,
                "Undefined variable '" + name.lexeme + "'.");
    }

}

File: slox/src/main/scala/ar/edu/utn/tadp/slox/environment/Environment.scala
package ar.edu.utn.tadp.slox.environment

import ar.edu.utn.tadp.slox.result.EvaluationResult
import ar.edu.utn.tadp.slox.tokens.Token
import ar.edu.utn.tadp.slox.values.Value

case class Environment(
    values: Map[String, Value] = Map.empty,
    enclosing: Option[Environment] = None
) {
  def define(name: String, value: Value): Environment =
    copy(values = values + (name -> value))

  def get(name: Token): EvaluationResult[Value] =
    values.get(name.lexeme) match
      case Some(value) => EvaluationResult.Success(value)
      case None =>
        enclosing match
          case Some(env) => env.get(name)
          case None =>
            EvaluationResult.RuntimeError(
              name,
              s"Undefined variable '${name.lexeme}'."
            )

  def getAt(distance: Int, name: String): Value =
    ancestor(distance).values(name)

  def assignAt(distance: Int, name: Token, value: Value): Environment =
    if distance == 0 then copy(values = values + (name.lexeme -> value))
    else
      val updatedEnclosing = enclosing.get.assignAt(distance - 1, name, value)
      copy(enclosing = Some(updatedEnclosing))

  def ancestor(distance: Int): Environment =
    if distance == 0 then this
    else enclosing.get.ancestor(distance - 1)

  def assign(name: Token, value: Value): EvaluationResult[(Value, Environment)] =
    if values.contains(name.lexeme) then
      EvaluationResult.Success(
        (value, copy(values = values + (name.lexeme -> value)))
      )
    else
      enclosing match
        case Some(env) =>
          env.assign(name, value).map { case (v, updatedEnclosing) =>
            (v, copy(enclosing = Some(updatedEnclosing)))
          }
        case None =>
          EvaluationResult.RuntimeError(
            name,
            s"Undefined variable '${name.lexeme}'."
          )
}
